query-engine/core/src/query_graph_builder/write/utils.rs

```diff
@@ -1014,3 +1017,217 @@ pub fn emulate_on_update_cascade(
+/// Foreign keys need to reference an existing record. Before the update is performed, we check that the new set of foreign keys points to an actual record.
+/// If they don't, we abort the query and error out. In the case of optional relations, the check is skipped if any of the foreign keys are set to `NULL`.
+///
+/// ```text
+///    ┌ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─    
+///         Parent (Read node)    │   
+/// ┌ ─│  (ids and foreign keys)   ─ ┐
+///     ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ┘   
+/// │                │               │
+///   ┌ ─ ─ ─ ─ ─ ─ ─▼─ ─ ─ ─ ─ ─ ─   
+/// │  ┌──────────────────────────┐│ │
+///   ││            If            │   
+/// │  │(no foreign keys are NULL)││ │
+///   │└──────────────────────────┘   
+/// │                │             │ │
+///   │            Then               
+/// │                │             │ │
+///   │              ▼                
+/// │  ┌──────────────────────────┐│ │
+///  ─▶│Find new referenced record│   
+///    └──────────────────────────┘│ │
+///   │              │                
+///                  │             │ │
+///   │              ▼                
+///    ┌──────────────────────────┐│ │
+///   ││ Check referenced record  │   
+///    │          exist           ││ │
+///   ││         (Empty)          │   
+///    └──────────────────────────┘│ │
+///   └ ─ ─ ─ ─ ─ ─ ─│─ ─ ─ ─ ─ ─ ─   
+///                  │               │
+///                  ▼                
+///       ┌ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─      │
+///               Update       │◀ ─ ─
+///       └ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─       
+/// ```
+#[allow(clippy::mutable_key_type)]
+pub fn preserve_referential_integrity_on_update(
+    graph: &mut QueryGraph,
+    connector_ctx: &ConnectorContext,
+    model_to_update: &ModelRef,
+    read_node: &NodeRef,
+    update_node: &NodeRef,
+) -> QueryGraphBuilderResult<()> {
+    // If the connector uses the `ReferentialIntegrity::ForeignKeys` mode, we do not do any checks / emulation.
+    if connector_ctx.referential_integrity.uses_foreign_keys() {
+        return Ok(());
+    }
+
+    let update_args = extract_update_args(graph.node_content(update_node).unwrap());
+    let relations = model_to_update.fields().relation();
+    let relations_to_update: Vec<_> = relations
+        .into_iter()
+        // We only pick required inlined relations. Updating the primary keys would trigger referential actions.
+        // Referential integrity only needs to be preserved when updating the foreign keys of the inlined side
+        .filter(|rf| rf.is_inlined_on_enclosing_model())
+        .filter_map(|rf| find_fks_to_be_updated(rf, update_args))
+        .collect();
+
+    for (rf, fks_to_be_updated) in relations_to_update {
+        let child_model = rf.related_model();
+        let child_model_identifier = rf.related_model().primary_identifier();
+        let fks = rf.scalar_fields();
+        let pks = rf.referenced_fields();
```

Comment:
Both `fks` and `pks` have misleading naming. I can somewhat live with `fks`, but `pks` is incorrect: It's the fields the FK refers to, which isn't necessarily the primary key. Usually my naming is along the lines of `relation_scalars` (or `rf_scalars`) and `referenced_scalars`.